File: api/src/main/java/mb/pie/api/TaskDef.java
package mb.pie.api;

import org.checkerframework.checker.nullness.qual.Nullable;

import java.io.Serializable;

/**
 * Definition of an incremental task which takes objects of type {@code I} and produces objects of type {@code O} when
 * executed.
 *
 * <p>
 * Inputs of type {@code I} must adhere to the following properties:
 * <ul>
 * <li>Implement {@link Serializable}</li>
 * <li>Implement {@link Object#equals(Object)} and {@link Object#hashCode()}</li>
 * <li>Must NOT be {@code null}</li>
 * <li>If the input is used as a key, it must also adhere to the properties of {@link #key(Serializable)} below</li>
 * </ul>
 * <p>
 * Furthermore, keys returned by the {@link #key(Serializable)} method, must adhere to the following properties:
 * <ul>
 * <li>Implement {@link Serializable}</li>
 * <li>Implement {@link Object#equals(Object)} and {@link Object#hashCode()}</li>
 * <li>Must NOT be {@code null}</li>
 * <li>
 * {@link Object#equals(Object)} and {@link Object#hashCode()} must return the same values after a serialization
 * roundtrip (e.g., serialize-deserialize)
 * </li>
 * <li>
 * The key's serialized bytes must be equal when the key's {@link Object#equals(Object)} method returns {@code true}
 * </li>
 * <p>
 * Finally, outputs of type {@code O} must adhere to the following properties:
 * <ul>
 * <li>Implement {@link Serializable}</li>
 * <li>Implement {@link Object#equals(Object)} and {@link Object#hashCode()}</li>
 * <p>
 * Failure to adhere to these properties will cause unsound incrementality.
 *
 * @param <I> Type of input objects. Must be {@link Serializable} and may NOT be {@code null}.
 * @param <O> Type of output objects. Must be {@link Serializable} but may be {@code null}.
 */
public interface TaskDef<I extends Serializable, O extends @Nullable Serializable> {
    /**
     * Gets the unique identifier of this task definition.
     *
     * @return Unique identifier of this task definition.
     */
    String getId();

    /**
     * Executes the task with given input, and returns its output.
     *
     * @param context Execution context for requiring tasks, and requiring/providing resources.
     * @param input   Input object.
     * @return Output object.
     * @throws Exception            When execution of the task fails unexpectedly.
     * @throws InterruptedException When execution of the task is cancelled or otherwise interrupted.
     */
    O exec(ExecContext context, I input) throws Exception;


    /**
     * Returns a key that uniquely identifies a task for given input. Defaults to entire input.
     *
     * @param input Input object.
     * @return Key for given input.
     */
    default Serializable key(I input) {
        return input;
    }

    /**
     * Returns the description of task for given {@code input}, with given {@code maxLength}. Defaults to {@code
     * "ID(input)"}.
     */
    default String desc(I input, int maxLength) {
        return this.getId() + '(' + StringUtil.toShortString(input.toString(), maxLength) + ')';
    }


    /**
     * Creates a {@link Task task instance} for this task definition with given {@code input}.
     */
    default Task<O> createTask(I input) {
        return new Task<>(this, input);
    }

    /**
     * Creates a {@link STask serializable task instance} for this task definition with given {@code input}.
     */
    default STask<O> createSerializableTask(I input) {
        return new STask<>(this, input);
    }

    /**
     * Creates a {@link STaskDef serializable task definition wrapper} for this task definition.
     */
    default STaskDef<I, O> createSerializableTaskDef() {
        return new STaskDef<>(this);
    }
}

File: api/src/main/kotlin/mb/pie/api/TaskDef.kt
package mb.pie.api

import java.io.Serializable

/**
 * Type for task inputs. It must adhere to the following properties:
 *
 * * Implements [Serializable].
 * * Implements [equals][Object.equals] and [hashCode][Object.hashCode].
 * * Must NOT be `null`.
 * * If they input is used as a [key][Key], it must also adhere to [key][Key]'s properties.
 *
 * Failure to adhere to these properties will cause unsound incrementality.
 */
typealias In = Serializable

/**
 * Type for task keys. It must adhere to the following properties:
 *
 * * Implements [Serializable].
 * * Implements [equals][Object.equals] and [hashCode][Object.hashCode].
 * * Must NOT be `null`.
 * * [Equals][Object.equals] and [hashCode][Object.hashCode] must return the same values after a serialization roundtrip (e.g., serialize-deserialize).
 * * The key's serialized bytes must be equal when the key's [equals][Object.equals] method returns true.
 *
 * Failure to adhere to these properties will cause unsound incrementality.
 */
typealias Key = Serializable

/**
 * Definition of an executable task.
 */
interface TaskDef<I : In, O : Out> {
  /**
   * Unique identifier of the task definition.
   */
  val id: String

  /**
   * Returns a key that uniquely identifies the task for given input.
   */
  @JvmDefault
  fun key(input: I): Key = input

  /**
   * Executes the task with given input, and returns its output.
   *
   * @throws ExecException when execution of the task fails unexpectedly.
   * @throws InterruptedException when execution of the task is cancelled.
   */
  @Throws(ExecException::class, InterruptedException::class)
  fun ExecContext.exec(input: I): O

  /**
   * Returns the description of task for given [input].
   */
  @JvmDefault
  fun desc(input: I): String = "$id($input)"

  /**
   * Returns the description of task for given [input], with given [maximum length][maxLength].
   */
  @JvmDefault
  fun desc(input: I, maxLength: Int = 100): String = "$id(${input.toString().toShortString(maxLength)})"

  /**
   * Creates a task instance with given [input] for this task definition.
   */
  @JvmDefault
  fun createTask(input: I): Task<I, O> = Task(this, input)
}

/**
 * [TaskDef] implementation using anonymous functions.
 */
open class LambdaTaskDef<I : In, O : Out>(
  override val id: String,
  private val execFunc: ExecContext.(I) -> O,
  private val keyFunc: ((I) -> Key)? = null,
  private val descFunc: ((I, Int) -> String)? = null
) : TaskDef<I, O> {
  override fun ExecContext.exec(input: I): O = execFunc(input)
  override fun key(input: I) = keyFunc?.invoke(input) ?: super.key(input)
  override fun desc(input: I, maxLength: Int): String = descFunc?.invoke(input, maxLength) ?: super.desc(input, maxLength)
}
